File: oolab/src/main/java/agh/ics/oop/model/Vector2d.java
package agh.ics.oop.model;

import java.util.Vector;

public class Vector2d {
    private final int x;
    private final int y;

    public Vector2d(int x, int y) {
        this.x = x;
        this.y = y;
    }

    public int getX() {
        return this.x;
    }
    public int getY() {
        return this.y;
    }

    public String toString() {
        return "(" + this.x + "," + this.y + ")";
    }

    public boolean precedes(Vector2d other) {
        return this.x < other.x && this.y <= other.y;
    }
    public boolean follows(Vector2d other) {
        return this.x > other.x && this.y >= other.y;
    }

    public Vector2d add(Vector2d other) {
        return new Vector2d(this.x + other.x, this.y + other.y);
    }

    public Vector2d subtract(Vector2d other) {
        return new Vector2d(this.x - other.x, this.y - other.y);
    }

    public Vector2d upperRight(Vector2d other) {
        int x_add = Math.max(this.x, other.x);
        int y_add = Math.max(this.y, other.y);
        return new Vector2d(x_add, y_add);
    }
    public Vector2d lowerLeft(Vector2d other) {
        int x_add = Math.min(this.x, other.x);
        int y_add = Math.min(this.y, other.y);
        return new Vector2d(x_add, y_add);
    }
    public Vector2d opposite(Vector2d other) {
        return new Vector2d(-other.x, -other.y);
    }
    public boolean equals(Vector2d other) { //TO TRZEBA JAKOS ZMIENIC
        return this.x == other.x && this.y == other.y;
    }

}

File: Kotlin/src/main/kotlin/model/Vector2d.kt
package model

import util.MapDirection

data class Vector2d(val x: Int, val y: Int) {

    override fun toString() = "($x, $y)"

    operator fun plus(other: Vector2d) = Vector2d(x + other.x, y + other.y)

    operator fun minus(other: Vector2d) = Vector2d(x - other.x, y - other.y)

    fun precedes(other: Vector2d) = x <= other.x && y <= other.y

    fun follows(other: Vector2d) = x >= other.x && y >= other.y

    fun upperRight(other: Vector2d) = Vector2d(maxOf(x, other.x), maxOf(y, other.y))

    fun lowerLeft(other: Vector2d) = Vector2d(minOf(x, other.x), minOf(y, other.y))

    fun opposite() = Vector2d(-x, -y)
}

fun MapDirection.toUnitVector(): Vector2d {
    return when (this) {
        MapDirection.NORTH -> Vector2d(0, 1)
        MapDirection.WEST -> Vector2d(-1, 0)
        MapDirection.SOUTH -> Vector2d(0, -1)
        MapDirection.EAST -> Vector2d(1, 0)
    }
}